integration-tests/src/main/java/org/apache/druid/testing/utils/MsqTestQueryHelper.java

```diff
@@ -0,0 +1,236 @@
+  }
+
+  /**
+   * Polls the overlord API every 1 second and waits for a submitted MSQ task to be completed. Alternatively, one can
+   * specify the maximum time to poll. The method returns the last fetched {@link TaskState} of the task
+   */
+  public TaskState pollTaskIdForCompletion(String taskId, long maxTimeoutSeconds)
+  {
+    if (maxTimeoutSeconds < 0) {
+      throw new IAE("Timeout cannot be negative");
+    } else if (maxTimeoutSeconds == 0) {
+      maxTimeoutSeconds = Long.MAX_VALUE;
+    }
+    long time = 0;
+    do {
+      TaskStatusPlus taskStatusPlus = overlordClient.getTaskStatus(taskId);
+      TaskState statusCode = taskStatusPlus.getStatusCode();
+      if (statusCode != null && statusCode.isComplete()) {
+        return taskStatusPlus.getStatusCode();
+      }
+      try {
+        Thread.sleep(1000);
+      }
+      catch (InterruptedException e) {
+        throw new ISE(e, "Interrupted while polling for task [%s] completion", taskId);
+      }
+    } while (time++ < maxTimeoutSeconds);
+    return overlordClient.getTaskStatus(taskId).getStatusCode();
+  }
+
+  /**
+   * Fetches status reports for a given task
+   */
+  public Map<String, MSQTaskReport> fetchStatusReports(String taskId)
+  {
+    return overlordClient.getTaskReportForMsqTask(taskId);
+  }
+
+  /**
+   * Compares the results for a given taskId. It is required that the task has produced some results that can be verified
+   */
+  public void compareResults(String taskId, MsqQueryWithResults expectedQueryWithResults)
+  {
+    Map<String, MSQTaskReport> statusReport = fetchStatusReports(taskId);
+    MSQTaskReport taskReport = statusReport.get(MSQTaskReport.REPORT_KEY);
+    if (taskReport == null) {
+      throw new ISE("Unable to fetch the status report for the task [%]", taskId);
+    }
+    MSQTaskReportPayload taskReportPayload = Preconditions.checkNotNull(
+        taskReport.getPayload(),
+        "payload"
+    );
+    MSQResultsReport resultsReport = Preconditions.checkNotNull(
+        taskReportPayload.getResults(),
+        "Results report for the task id is empty"
+    );
+
+    List<Map<String, Object>> actualResults = new ArrayList<>();
+
+    Yielder<Object[]> yielder = resultsReport.getResultYielder();
+    RowSignature rowSignature = resultsReport.getSignature();
+
+    while (!yielder.isDone()) {
+      Object[] row = yielder.get();
+      Map<String, Object> rowWithFieldNames = new LinkedHashMap<>();
+      for (int i = 0; i < row.length; ++i) {
+        rowWithFieldNames.put(rowSignature.getColumnName(i), row[i]);
+      }
+      actualResults.add(rowWithFieldNames);
+      yielder = yielder.next(null);
+    }
+
+    boolean resultsComparison = QueryResultVerifier.compareResults(
+        actualResults,
+        expectedQueryWithResults.getExpectedResults(),
+        Collections.emptyList()
+    );
+    if (!resultsComparison) {
+      throw new IAE("Expected query result is different from the actual result");
```

Comment:
I know that `AbstractTestQueryHelper` also does this in a similar way but we should avoid this if we want to write cleaner tests. Assertions should preferably be done using the appropriate JUnit/TestNG constructs.

Reply:
Maybe `QueryResultVerifier` itself should do the assertions or atleast give back messages detailing the mismatch. The caller getting a boolean and then throwing an ISE makes for very difficult to debug tests.

Reply:
I second Kashif's opinion. The signature of the resultVerifier can be "Optional<StringErrorMessage>" and then the caller can do whatever it wants. 
Though one version in the results verifier should always throw an exception. 

void resultsEqualsOrThrow()

Something like that.